source/common/upstream/load_balancer_impl.h

```diff
@@ -138,21 +196,27 @@ class ZoneAwareLoadBalancerBase : public LoadBalancerBase {
 
 /**
  * Implementation of LoadBalancer that performs RR selection across the hosts in the cluster.
+ * This scheduler respects host weighting and utilizes an EdfScheduler to achieve O(log n)
```

Comment:
Other than Wikipedia are there any good links for background reading on this approach? (I might at least provide the EDS Wikipedia link).

Reply:
Not sure there is a great resource here. I started by looking at http://kb.linuxvirtualserver.org/wiki/Weighted_Round-Robin_Scheduling. This algorithm seemed to be O(n) pick, which wasn't good. There is the option of precomputing the schedule,
which seemed to be O(m * n) space and construction time where m is the weight range with O(1) pick, so this won't work if we scale the number of weights (but is probably fine if we stick with 0-128 forever). There's probably reasonable heuristic approximation as well.

We use this EDF schedule approach at Google, so it seemed a reasonable compromise with O(log n) pick and O(n) space utilization. I tweaked it by adding the `order_offset_` to ensure we get the FIFO RR behavior independent of what `std::priority_queue` is doing under the hood when we queue up at the same priority in between picks.

LMK if you have suggestions on how to better document, I'll add links to basic EDF. We could always go with explicit queue construction if you feel that we will never grow the priority range, as 128 seems a reasonably low constant factor.

Reply:
FWIW, the other advantage to a precomputed schedule would be that weighted RR and unweighted are effectively the same when weights are identical.

One other useful place for the EDF is when we want to later support dynamic weighting and use things like upstream host load factors, which might be floats; while we could approximate by rounding to some discrete set of integers (e.g. [0, 128)), we might prefer to just allow arbitrary precision.

LMK what you think here, I'm going to hold back on other changes until we decide on this.

Reply:
If this is how it's done at Google and it is working out this seems reasonable to me. I guess my only ask would be to maybe put all of the above into the comment for the load balancer just to let the reader better understand why we are doing what we doing?